### PR TITLE
fix: Cancel query in local mode.

### LINKED ISSUE #7285: bug(batch): Ctrlc + C doesn't work in local mode.
### Describe the bug

When user issued a complex query in local mode, it may run for a long time and user want to cancel it, but currently `ctrl + c` is working for query in local mode. 

### To Reproduce

Issue a complex query in local mode.

### Expected behavior

It should stop.

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by liurenjie1024**
I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

## What's changed and what's your intention?

Supports cancel query in local mode.

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features).
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

### Types of user-facing changes

Please keep the types that apply to your changes, and remove those that do not apply.

* Other (please specify in the release note below)

### Release note

When a query runs in local mode, now user can use `ctrl + c` to cancel it.

## Refer to a related PR or issue link (optional)

Closes #7285

### CODE REVIEW

File: src/frontend/src/session.rs [hunk 1]
```diff
@@ -471,6 +483,31 @@ impl SessionImpl {
...
+    }
+
+    pub fn reset_cancel_query_flag(&self) -> Tripwire<std::result::Result<DataChunk, BoxedError>> {
+        let mut flag = self.current_query_cancel_flag.write();
+        let (trigger, tripwire) = stream_tripwire(|| Err(Box::new(QueryCancelError) as BoxedError));
+        *flag = Some(trigger);
+        tripwire
+    }
+
+    pub fn cancel_current_query(&self) {
```
Comment (BowenXiao1999) [inline src/frontend/src/session.rs:497]: Is it will cancel all local query in one session?

File: src/common/src/util/stream_cancel.rs [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+    pub fn abort(self) {
+        if self.sender.send(()).is_err() {
+            tracing::debug!("Trigger receiver closed, ignoring...");
+        }
+    }
+}
+
+pub struct Tripwire<R> {
+    receiver: Receiver<()>,
+    cancel_result: Box<dyn FnOnce() -> R + Send>,
```
Comment (BowenXiao1999) [inline src/common/src/util/stream_cancel.rs:43]: Is it the cancel success result? Maybe we just need to log it

File: src/frontend/src/session.rs (hunk 1 above)
Comment (liurenjie1024) [reply src/frontend/src/session.rs:497]: No, only current running query

File: src/common/src/util/stream_cancel.rs (hunk 2 above)
Comment (liurenjie1024) [reply src/common/src/util/stream_cancel.rs:43]: No, it's used to send cancel error.

File: src/common/src/util/stream_cancel.rs [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+    (
+        Trigger { sender },
+        Tripwire {
+            receiver,
+            cancel_result: Box::new(cancel_result),
+        },
+    )
+}
+
+pub fn cancellable_stream<S>(
```
Comment (BugenZhao) [inline src/common/src/util/stream_cancel.rs:60]: There's an `Abortable` in `futures`. Does it meet our requirements?

Comment (liurenjie1024) [reply src/common/src/util/stream_cancel.rs:60]: No, I want to user to receive query cancel error when query cancelled.

Comment (BugenZhao) [reply src/common/src/util/stream_cancel.rs:60]: `Abortable` will wrap the original stream and yields `Err(Aborted)` after cancelling.

https://docs.rs/futures/latest/futures/stream/struct.Abortable.html#impl-Abortable%3CT%3E

Comment (liurenjie1024) [reply src/common/src/util/stream_cancel.rs:60]: It just stop stream

Comment (BugenZhao) [reply src/common/src/util/stream_cancel.rs:60]: Don't quite get that. Doesn't `take_until` also stop the stream? If you mean doing some "post work" after it's aborted, simply map the `Err` then we can get it done there.

Comment (liurenjie1024) [reply src/common/src/util/stream_cancel.rs:60]: I mean when we abort stream, no error returned

Comment (BugenZhao) [reply src/common/src/util/stream_cancel.rs:60]: It's just the interface design used as a signal of abortion of `futures::Abortable`. 😂 If we don't want the `Err` itself, transform or discard it before turning the stream into an opaque type.

Comment (liurenjie1024) [reply src/common/src/util/stream_cancel.rs:60]: I think it's hard to design the api.

File: src/common/src/util/stream_cancel.rs [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+            } else {
+                polled = true;
+                if s_cancel_flag.load(Ordering::SeqCst) {
+                    Poll::Ready(Some(cancel_result.take().unwrap()()))
+                } else {
+                    Poll::Ready(None)
+                }
+            }
+        })
+    };
```
Comment (BugenZhao) [inline src/common/src/util/stream_cancel.rs:90]: Does `once` work here?

```rust
    let cancel_stream = once(async move {
        if s_cancel_flag.load(Ordering::SeqCst) {
            Some(cancel_result())
        } else {
            None
        }
    })
    .map(futures::stream::iter)
    .flatten();
```

File: src/frontend/src/session.rs [hunk 5]
```diff
@@ -357,6 +360,11 @@ pub struct SessionImpl {
 
     /// Identified by process_id, secret_key. Corresponds to SessionManager.
     id: (i32, i32),
+
+    /// Query cancel flag.
+    /// This flag is set only when current query is executed in local mode, and used to cancel
+    /// local query.
+    current_query_cancel_flag: RwLock<Option<Trigger>>,
```
Comment (BugenZhao) [inline src/frontend/src/session.rs:367]: Seems there's no calling of `read`.

File: src/common/src/util/stream_cancel.rs [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+        if receiver.await.is_err() {
+            tracing::debug!("Cancel sender closed, ignoring...");
+        } else {
+            cancel_flag.store(true, Ordering::SeqCst);
+        }
+    };
+
+    let data_stream = data_stream.take_until(cancel_future).chain(cancel_stream);
+
+    Box::pin(data_stream)
```
Comment (BugenZhao) [inline src/common/src/util/stream_cancel.rs:102]: As a general component, I think we may not `box` here and let the caller to decide that.

File: src/common/src/util/stream_cancel.rs [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+impl Trigger {
+    /// Used to trigger cancelling of stream.
+    pub fn abort(self) {
+        if self.sender.send(()).is_err() {
+            tracing::debug!("Trigger receiver closed, ignoring...");
+        }
+    }
+}
+
+pub struct Tripwire<R> {
```
Comment (BugenZhao) [inline src/common/src/util/stream_cancel.rs:41]: Could you please add some rustdocs for these public stuff?

**Review by BugenZhao — COMMENTED** [review]
Rest LGTM!

File: src/common/src/util/stream_cancel.rs (hunk 7 above)
Comment (liurenjie1024) [reply src/common/src/util/stream_cancel.rs:59]: I think it's better to add an example to doc.

**Review by BowenXiao1999 — APPROVED** [review]

**Review by BugenZhao — APPROVED** [review]
LGTM

**Pull request merged**